features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/lifecycle_api_router.jag

```diff
@@ -131,364 +131,405 @@ require('/modules/publisher.js').exec(function (ctx) {
-				return;
-			}
-
-		}
-
-		// everything went great
-		response.status = 200;
-		response.contentType = "application/json";
-		content.status = "Success";
-		response.content = stringify(content);
-
-	});
-
-	function promoteArtifactLifeCycle(artifactManager, artifact, artifactType, action) {
-		var result = {};
-		try {
-			artifactManager.promoteLifecycleState(action, artifact);
-			result.isSuccess = true;
-			result.message = getMessageForLifeCyclePromotion(artifact, artifactType, action, true);
-		} catch (e) {
-			// GovernanceArtifactImpl.invokeAction(action) throws GovernanceException
-			result.isSuccess = false;
-			result.message = getMessageForLifeCyclePromotion(artifact, artifactType, action, false);
-			log.error(result.message);
-			log.error(e);
-		}
-		return result;
-	}
-
-	function getMessageForLifeCyclePromotion(artifact, artifactType, action, isSuccess) {
-		// actions (from '/config/publisher.json' --> publisherActions):
+    /**
+     * API for handling web/mobile apps lifecycle transitions. Content type of the PUT request
+     * should be 'application/json'.
+     * @returns {{status: string, messages: string[], artifacts: Object[]}}
+     */
+    routeManager.register('PUT', 'publisher', '/publisher/api/lifecycle/{action}/{type}/{id}', function (context) {
+
+        var appType = context.params.type; // webapp or mobileapp
+        var action = context.params.action; // see '/config/publisher.json' --> publisherActions
+        var uuid = context.params.id; // UUID
+
+        /** @type {{status: string, messages: string[], artifacts: Object[]}} */
+        var content = {status: "", messages: [], artifacts: []};
+
+        // validating action
+        if (config.publisherActions.indexOf(action) == -1) {
+            log.error("Invalid action '" + action + "' invoked on a " + appType
+                      + " with UUID " + uuid);
+            response.status = 422; // Unprocessable Entity (WebDAV; RFC 4918)
+            response.contentType = "application/json";
+            content.status = "Invalid Action";
+            content.messages.push("Action '" + action + "' cannot be performed on "
+                                  + appType + "s");
+            response.content = stringify(content);
+            return;
+        }
+
+        var userManager = server.userManager(server.current(session).tenantId);
+        var user = server.current(session);
+        var permission = null;
+        if (appType == 'webapp') {
+            permission = config.permissions.webapp_publish;
+        } else if (appType == 'mobileapp') {
+            permission = config.permissions.mobileapp_publish;
+        }
+        // check whether current user has publisher permissions
+        var hasPublisherPermission = permissions.isAuthorized(user.username, permission, userManager);
+        if (!hasPublisherPermission) {
+            log.error("Insufficient permissions to perform '" + action + "' on " + appType);
```

Comment:
messages formatted, see commit 42edb3b60ee62d4da43658986e3fb68d2b0a61e4 of PR #185